File: tests/python/test_starrocks.py
from __future__ import annotations

import datetime as dt
import unittest
from decimal import Decimal

from _support import ConnectionManager, ColumnType, QualifiedNamePartRole, assert_boolean_value, assert_column_metadata, assert_namespace_access, assert_non_empty_value, assert_table_qualified_name, assert_type_coverage, execute_non_query, find_result_set_row_index, is_runtime_unavailable_error, load_test_target, read_result_set_values, should_run_section, unique_identifier


STARROCKS_ADDITIONAL_TYPES_SQL = """\
select
    cast(1 as tinyint) as tiny_value,
    cast(2 as smallint) as small_value,
    cast(3 as int) as int_value,
    cast(4 as bigint) as big_value,
    cast(5.5 as float) as float_value,
    cast(6.5 as double) as double_value,
    cast('[1,2,3]' as array<int>) as array_value,
    map('a',1,'b',2) as map_value,
    row(1, 'alpha') as struct_value
"""

STARROCKS_SKETCH_TYPES_SQL = """\
select
    to_bitmap(42) as bitmap_value,
    hll_hash('alpha') as hll_value,
    percentile_hash(42) as percentile_value
"""


def build_starrocks_type_coverage_case(table_name: str) -> dict[str, object]:
    return {
        "metadata_database": None,
        "create_table_sql": f"""\
create table {table_name} (
    id bigint not null,
    bool_value boolean not null,
    int_value bigint not null,
    float_value double not null,
    text_value string not null,
    fixed_text_value char(5) not null,
    decimal_value decimal(10, 2) not null,
    date_value date not null,
    timestamp_value datetime not null,
    largeint_value largeint not null,
    json_value json not null
) duplicate key(id) distributed by hash(id) buckets 1
properties ("replication_num" = "1")""",
        "insert_sql": f"""\
insert into {table_name} (
    id, bool_value, int_value, float_value, text_value, fixed_text_value, decimal_value, date_value,
    timestamp_value, largeint_value, json_value
) values (
    1, true, 42, 3.5, 'alpha', 'omega', 123.45, '2024-01-02', '2024-01-02 03:04:05',
    123456789012345678901234567890, parse_json('{"enabled": true, "count": 1}')
)""",
        "select_sql": f"""\
select id, bool_value, int_value, float_value, text_value, fixed_text_value,
    decimal_value, date_value, timestamp_value, largeint_value, json_value
from {table_name}
order by id""",
        "expected_columns": [
            {"name": "id", "column_type": ColumnType.INT64},
            {"name": "bool_value", "column_type": [ColumnType.BOOLEAN, ColumnType.INT8]},
            {"name": "int_value", "column_type": ColumnType.INT64},
            {"name": "float_value", "column_type": ColumnType.FLOAT64},
            {"name": "text_value", "column_type": ColumnType.TEXT},
            {"name": "fixed_text_value", "column_type": ColumnType.TEXT},
            {"name": "decimal_value", "column_type": ColumnType.DECIMAL},
            {"name": "date_value", "column_type": ColumnType.DATE},
            {"name": "timestamp_value", "column_type": ColumnType.TIMESTAMP},
            {"name": "largeint_value", "column_type": ColumnType.TEXT},
            {"name": "json_value", "column_type": [ColumnType.JSON, ColumnType.TEXT]},
        ],
    }


def assert_starrocks_type_coverage_values(result_set: object) -> None:
    columns = result_set.columns
    assert result_set.value(0, 0) == 1

    bool_value = result_set.value(0, 1)
    if columns[1].column_type == ColumnType.BOOLEAN:
        assert_boolean_value(bool_value)
    else:
        assert bool_value == 1

    assert result_set.value(0, 2) == 42
    assert result_set.value(0, 3) == 3.5
    assert result_set.value(0, 4) == "alpha"
    assert result_set.value(0, 5) == "omega"
    assert result_set.value(0, 6) == Decimal("123.45")
    assert result_set.value(0, 7) == dt.date(2024, 1, 2)
    timestamp_value = result_set.value(0, 8)
    assert timestamp_value == dt.datetime(2024, 1, 2, 3, 4, 5)
    assert result_set.value(0, 9) == "123456789012345678901234567890"

    json_value = result_set.value(0, 10)
    if columns[10].column_type == ColumnType.JSON:
        assert json_value == {"enabled": True, "count": 1}
    else:
        assert isinstance(json_value, str)
        assert '"enabled"' in json_value


async def assert_starrocks_additional_type_coverage(connection: object) -> None:
    result_set = await connection.execute_async(STARROCKS_ADDITIONAL_TYPES_SQL)
    try:
        assert_column_metadata(result_set.columns, [
            {"name": "tiny_value", "column_type": ColumnType.INT8},
            {"name": "small_value", "column_type": ColumnType.INT16},
            {"name": "int_value", "column_type": ColumnType.INT32},
            {"name": "big_value", "column_type": ColumnType.INT64},
            {"name": "float_value", "column_type": ColumnType.FLOAT32},
            {"name": "double_value", "column_type": ColumnType.FLOAT64},
            {"name": "array_value", "column_type": ColumnType.TEXT},
            {"name": "map_value", "column_type": ColumnType.TEXT},
            {"name": "struct_value", "column_type": ColumnType.TEXT},
        ])
        assert all(column.raw_type is None for column in result_set.columns)
        assert result_set.value(0, 0) == 1
        assert result_set.value(0, 1) == 2
        assert result_set.value(0, 2) == 3
        assert result_set.value(0, 3) == 4
        assert result_set.value(0, 4) == 5.5
        assert result_set.value(0, 5) == 6.5
        assert result_set.value(0, 6) == "[1,2,3]"
        assert result_set.value(0, 7) == '{"a":1,"b":2}'
        assert result_set.value(0, 8) == '{"col1":1,"col2":"alpha"}'
    finally:
        await result_set.close_async()

    result_set = await connection.execute_async(STARROCKS_SKETCH_TYPES_SQL)
    try:
        assert_column_metadata(result_set.columns, [
            {"name": "bitmap_value", "column_type": ColumnType.TEXT},
            {"name": "hll_value", "column_type": ColumnType.TEXT},
            {"name": "percentile_value", "column_type": ColumnType.BINARY},
        ])
        assert result_set.value(0, 0) is None
        assert result_set.value(0, 1) is None
        assert result_set.value(0, 2) is None
    finally:
        await result_set.close_async()


class StarRocksBindingIntegrationTest(unittest.IsolatedAsyncioTestCase):
    async def test_starrocks(self) -> None:
        section = "starrocks"
        if not should_run_section(section):
            self.skipTest(f"DATABASE_ZIG_TEST_SECTION is filtering out {section}")

        target = load_test_target(section)
        database_name = unique_identifier("aq_sr")
        table_name = unique_identifier("records")
        missing_database = unique_identifier("missing_db")

        async with ConnectionManager() as manager:
            try:
                admin_connection = await manager.connect_async(target.driver, target.dsn())
            except RuntimeError as error:
                if is_runtime_unavailable_error(error):
                    self.skipTest(str(error))
                raise

            with self.assertRaises(RuntimeError):
                await manager.connect_async(target.driver, target.dsn(missing_database))
            try:
                await execute_non_query(admin_connection, f"create database if not exists {database_name}")

                database_connection = await manager.connect_async(target.driver, target.dsn(database_name))
                try:
                    self.assertTrue(await database_connection.test_async())

                    await assert_type_coverage(database_connection, build_starrocks_type_coverage_case(table_name), assert_starrocks_type_coverage_values)
                    await assert_starrocks_additional_type_coverage(database_connection)

                    missing_table = unique_identifier("missing")
                    with self.assertRaisesRegex(RuntimeError, missing_table):
                        await database_connection.execute_async(f"select * from {missing_table}")

                    missing_column = unique_identifier("missing_column")
                    with self.assertRaisesRegex(RuntimeError, missing_column):
                        await database_connection.execute_async(f"select {missing_column} from {table_name}")

                    databases_result = await database_connection.get_databases_async()
                    try:
                        self.assertIn(database_name, read_result_set_values(databases_result, 0))
                    finally:
                        await databases_result.close_async()

                    with self.assertRaisesRegex(RuntimeError, "get catalogs is not supported"):
                        await database_connection.get_catalogs_async()

                    tables_result = await database_connection.get_tables_async(database=database_name)
                    try:
                        self.assertIn(table_name, read_result_set_values(tables_result, 2))
                        catalog_names = read_result_set_values(tables_result, 0)
                        self.assertTrue(any(name not in (None, "") for name in catalog_names))
                        assert_table_qualified_name(tables_result, find_result_set_row_index(tables_result, 2, table_name))
                    finally:
                        await tables_result.close_async()

                    namespace_access = await database_connection.inspect_namespace_access_async(database=database_name)
                    assert_namespace_access(
                        namespace_access,
                        can_get_schema=False,
                        has_catalog_access=True,
                        has_namespace_access=True,
                        namespace_role=QualifiedNamePartRole.DATABASE,
                        expected_parts=[(QualifiedNamePartRole.DATABASE, database_name)],
                    )

                    missing_access = await database_connection.inspect_namespace_access_async(database=missing_database)
                    assert_namespace_access(
                        missing_access,
                        can_get_schema=False,
                        has_catalog_access=True,
                        has_namespace_access=False,
                        namespace_role=QualifiedNamePartRole.DATABASE,
                        expected_parts=[(QualifiedNamePartRole.DATABASE, missing_database)],
                    )
                finally:
                    await database_connection.close_async()
            finally:
                try:
                    await execute_non_query(admin_connection, f"drop database if exists {database_name}")
                finally:
                    await admin_connection.close_async()


if __name__ == "__main__":
    unittest.main()

File: tests/rust/test_starrocks.rs
mod support;

use aq_database::{ColumnType, Value};
use support::{
    assert_columns_match, assert_namespace_access, assert_table_qualified_name, execute_non_query,
    find_row_index, load_target, maybe_skip_configured_section, read_values, ExpectedColumn,
    unique_identifier,
};

const STARROCKS_BOOL_TYPES: &[ColumnType] = &[ColumnType::Boolean, ColumnType::Int8];
const STARROCKS_JSON_TYPES: &[ColumnType] = &[ColumnType::Json, ColumnType::Text];

const STARROCKS_ADDITIONAL_TYPES_SQL: &str = concat!(
    "select ",
    "cast(1 as tinyint) as tiny_value, ",
    "cast(2 as smallint) as small_value, ",
    "cast(3 as int) as int_value, ",
    "cast(4 as bigint) as big_value, ",
    "cast(5.5 as float) as float_value, ",
    "cast(6.5 as double) as double_value, ",
    "cast('[1,2,3]' as array<int>) as array_value, ",
    "map('a',1,'b',2) as map_value, ",
    "row(1, 'alpha') as struct_value"
);

const STARROCKS_SKETCH_TYPES_SQL: &str = concat!(
    "select ",
    "to_bitmap(42) as bitmap_value, ",
    "hll_hash('alpha') as hll_value, ",
    "percentile_hash(42) as percentile_value"
);

#[test]
fn rust_binding_starrocks_lifecycle() {
    if let Some(reason) = maybe_skip_configured_section("starrocks") {
        eprintln!("skipping rust StarRocks integration test: {reason}");
        return;
    }

    let target = load_target("starrocks").expect("starrocks target should load after skip check");
    let database_name = unique_identifier("aq_sr");
    let table_name = unique_identifier("records");
    let missing_database = unique_identifier("missing_db");

    let test_result = (|| -> Result<(), Box<dyn std::error::Error>> {
        let manager = aq_database::ConnectionManager::new()?;

        let admin_connection = match manager.connect(target.driver, &target.dsn(None)) {
            Ok(connection) => connection,
            Err(error) if is_runtime_unavailable(&error) => {
                eprintln!("skipping rust StarRocks integration test at runtime: {error}");
                return Ok(());
            }
            Err(error) => return Err(Box::new(error)),
        };

        let missing_database_error = match manager.connect(target.driver, &target.dsn(Some(&missing_database))) {
            Ok(_) => panic!("missing database should fail"),
            Err(error) => error,
        };
        assert!(!missing_database_error.to_string().is_empty());
        execute_non_query(&admin_connection, &format!("create database if not exists {database_name}"))?;

        let database_result = (|| -> Result<(), Box<dyn std::error::Error>> {
            let database_connection = manager.connect(target.driver, &target.dsn(Some(&database_name)))?;
            assert!(database_connection.test()?);

            assert_starrocks_type_coverage(&database_connection, &table_name)?;
            assert_starrocks_additional_type_coverage(&database_connection)?;

            let missing_table = unique_identifier("missing");
            let missing_table_error = match database_connection.execute(&format!("select * from {missing_table}")) {
                Ok(_) => panic!("missing table should fail"),
                Err(error) => error,
            };
            assert!(!missing_table_error.to_string().is_empty());

            let missing_column = unique_identifier("missing_column");
            let missing_column_error = match database_connection.execute(&format!("select {missing_column} from {table_name}")) {
                Ok(_) => panic!("missing column should fail"),
                Err(error) => error,
            };
            assert!(!missing_column_error.to_string().is_empty());

            let databases_result = database_connection.get_databases()?;
            let database_names = read_values(&databases_result, 0)?;
            assert!(database_names.contains(&Value::Text(database_name.clone())));
            databases_result.close()?;

            let catalogs_error = match database_connection.get_catalogs() {
                Ok(_) => panic!("starrocks/mysql-compatible targets should reject get_catalogs"),
                Err(error) => error,
            };
            assert!(catalogs_error.to_string().contains("get catalogs is not supported"));

            let tables_result = database_connection.get_tables(None, Some(&database_name))?;
            let table_names = read_values(&tables_result, 2)?;
            assert!(table_names.contains(&Value::Text(table_name.clone())));
            let catalog_names = read_values(&tables_result, 0)?;
            assert!(catalog_names.iter().any(|value| match value {
                Value::Text(text) => !text.is_empty(),
                _ => false,
            }));
            let table_row = find_row_index(&tables_result, 2, &Value::Text(table_name.clone()))?;
            assert_table_qualified_name(&tables_result, table_row)?;
            tables_result.close()?;

            let namespace_access = database_connection.inspect_namespace_access(None, Some(&database_name))?;
            assert_namespace_access(
                &namespace_access,
                false,
                true,
                true,
                aq_database::QualifiedNamePartRole::Database,
                &[(aq_database::QualifiedNamePartRole::Database, &database_name)],
            );

            let missing_access = database_connection.inspect_namespace_access(None, Some(&missing_database))?;
            assert_namespace_access(
                &missing_access,
                false,
                true,
                false,
                aq_database::QualifiedNamePartRole::Database,
                &[(aq_database::QualifiedNamePartRole::Database, &missing_database)],
            );

            database_connection.close()?;
            Ok(())
        })();

        let drop_result = execute_non_query(&admin_connection, &format!("drop database if exists {database_name}"));
        let close_result = admin_connection.close();

        database_result?;
        drop_result?;
        close_result?;
        Ok(())
    })();

    test_result.unwrap();
}

fn assert_starrocks_type_coverage(
    connection: &aq_database::Connection,
    table_name: &str,
) -> Result<(), Box<dyn std::error::Error>> {
    execute_non_query(
        connection,
        &format!(
            "create table {table_name} (id bigint not null, bool_value boolean not null, int_value bigint not null, float_value double not null, text_value string not null, fixed_text_value char(5) not null, decimal_value decimal(10, 2) not null, date_value date not null, timestamp_value datetime not null, largeint_value largeint not null, json_value json not null) duplicate key(id) distributed by hash(id) buckets 1 properties (\"replication_num\" = \"1\")"
        ),
    )?;

    execute_non_query(
        connection,
        &format!(
            "insert into {table_name} (id, bool_value, int_value, float_value, text_value, fixed_text_value, decimal_value, date_value, timestamp_value, largeint_value, json_value) values (1, true, 42, 3.5, 'alpha', 'omega', 123.45, '2024-01-02', '2024-01-02 03:04:05', 123456789012345678901234567890, parse_json('{{\"enabled\": true, \"count\": 1}}'))"
        ),
    )?;

    let result_set = connection.execute(&format!(
        "select id, bool_value, int_value, float_value, text_value, fixed_text_value, decimal_value, date_value, timestamp_value, largeint_value, json_value from {table_name} order by id"
    ))?;
    assert_eq!(result_set.row_count()?, 1);
    let columns = result_set.columns()?;
    assert_columns_match(
        &columns,
        &[
            ExpectedColumn { name: "id", column_types: &[ColumnType::Int64], raw_type: None },
            ExpectedColumn { name: "bool_value", column_types: STARROCKS_BOOL_TYPES, raw_type: None },
            ExpectedColumn { name: "int_value", column_types: &[ColumnType::Int64], raw_type: None },
            ExpectedColumn { name: "float_value", column_types: &[ColumnType::Float64], raw_type: None },
            ExpectedColumn { name: "text_value", column_types: &[ColumnType::Text], raw_type: None },
            ExpectedColumn { name: "fixed_text_value", column_types: &[ColumnType::Text], raw_type: None },
            ExpectedColumn { name: "decimal_value", column_types: &[ColumnType::Decimal], raw_type: None },
            ExpectedColumn { name: "date_value", column_types: &[ColumnType::Date], raw_type: None },
            ExpectedColumn { name: "timestamp_value", column_types: &[ColumnType::Timestamp], raw_type: None },
            ExpectedColumn { name: "largeint_value", column_types: &[ColumnType::Text], raw_type: None },
            ExpectedColumn { name: "json_value", column_types: STARROCKS_JSON_TYPES, raw_type: None },
        ],
    );
    assert_eq!(result_set.value(0, 0)?, Value::Int64(1));
    match columns[1].column_type {
        ColumnType::Boolean => assert_eq!(result_set.value(0, 1)?, Value::Boolean(true)),
        _ => assert_eq!(result_set.value(0, 1)?, Value::Int8(1)),
    }
    assert_eq!(result_set.value(0, 2)?, Value::Int64(42));
    assert_eq!(result_set.value(0, 3)?, Value::Float64(3.5));
    assert_eq!(result_set.value(0, 4)?, Value::Text("alpha".to_owned()));
    assert_eq!(result_set.value(0, 5)?, Value::Text("omega".to_owned()));
    assert_eq!(result_set.value(0, 6)?, Value::Text("123.45".to_owned()));
    assert_eq!(result_set.value(0, 7)?, Value::Text("2024-01-02".to_owned()));
    assert_matches_text_prefix(result_set.value(0, 8)?, "2024-01-02T03:04:05");
    assert_eq!(
        result_set.value(0, 9)?,
        Value::Text("123456789012345678901234567890".to_owned())
    );
    match columns[10].column_type {
        ColumnType::Json => assert_eq!(result_set.value(0, 10)?, Value::Text("{\"enabled\": true, \"count\": 1}".to_owned())),
        _ => assert_contains_text(result_set.value(0, 10)?, "\"enabled\""),
    }
    result_set.close()?;

    let cursor = connection.cursor(&format!(
        "select id, bool_value, int_value, float_value, text_value, fixed_text_value, decimal_value, date_value, timestamp_value, largeint_value, json_value from {table_name} order by id"
    ))?;
    assert_columns_match(
        &cursor.columns()?,
        &[
            ExpectedColumn { name: "id", column_types: &[ColumnType::Int64], raw_type: None },
            ExpectedColumn { name: "bool_value", column_types: STARROCKS_BOOL_TYPES, raw_type: None },
            ExpectedColumn { name: "int_value", column_types: &[ColumnType::Int64], raw_type: None },
            ExpectedColumn { name: "float_value", column_types: &[ColumnType::Float64], raw_type: None },
            ExpectedColumn { name: "text_value", column_types: &[ColumnType::Text], raw_type: None },
            ExpectedColumn { name: "fixed_text_value", column_types: &[ColumnType::Text], raw_type: None },
            ExpectedColumn { name: "decimal_value", column_types: &[ColumnType::Decimal], raw_type: None },
            ExpectedColumn { name: "date_value", column_types: &[ColumnType::Date], raw_type: None },
            ExpectedColumn { name: "timestamp_value", column_types: &[ColumnType::Timestamp], raw_type: None },
            ExpectedColumn { name: "largeint_value", column_types: &[ColumnType::Text], raw_type: None },
            ExpectedColumn { name: "json_value", column_types: STARROCKS_JSON_TYPES, raw_type: None },
        ],
    );
    assert!(cursor.next()?);
    assert!(!cursor.next()?);
    cursor.close()?;

    Ok(())
}

fn assert_starrocks_additional_type_coverage(
    connection: &aq_database::Connection,
) -> Result<(), Box<dyn std::error::Error>> {
    let result_set = connection.execute(STARROCKS_ADDITIONAL_TYPES_SQL)?;
    assert_columns_match(
        &result_set.columns()?,
        &[
            ExpectedColumn { name: "tiny_value", column_types: &[ColumnType::Int8], raw_type: None },
            ExpectedColumn { name: "small_value", column_types: &[ColumnType::Int16], raw_type: None },
            ExpectedColumn { name: "int_value", column_types: &[ColumnType::Int32], raw_type: None },
            ExpectedColumn { name: "big_value", column_types: &[ColumnType::Int64], raw_type: None },
            ExpectedColumn { name: "float_value", column_types: &[ColumnType::Float32], raw_type: None },
            ExpectedColumn { name: "double_value", column_types: &[ColumnType::Float64], raw_type: None },
            ExpectedColumn { name: "array_value", column_types: &[ColumnType::Text], raw_type: None },
            ExpectedColumn { name: "map_value", column_types: &[ColumnType::Text], raw_type: None },
            ExpectedColumn { name: "struct_value", column_types: &[ColumnType::Text], raw_type: None },
        ],
    );
    assert_eq!(result_set.value(0, 0)?, Value::Int8(1));
    assert_eq!(result_set.value(0, 1)?, Value::Int16(2));
    assert_eq!(result_set.value(0, 2)?, Value::Int32(3));
    assert_eq!(result_set.value(0, 3)?, Value::Int64(4));
    assert_eq!(result_set.value(0, 4)?, Value::Float32(5.5));
    assert_eq!(result_set.value(0, 5)?, Value::Float64(6.5));
    assert_eq!(result_set.value(0, 6)?, Value::Text("[1,2,3]".to_owned()));
    assert_eq!(result_set.value(0, 7)?, Value::Text("{\"a\":1,\"b\":2}".to_owned()));
    assert_eq!(result_set.value(0, 8)?, Value::Text("{\"col1\":1,\"col2\":\"alpha\"}".to_owned()));
    result_set.close()?;

    let sketch_result = connection.execute(STARROCKS_SKETCH_TYPES_SQL)?;
    assert_columns_match(
        &sketch_result.columns()?,
        &[
            ExpectedColumn { name: "bitmap_value", column_types: &[ColumnType::Text], raw_type: None },
            ExpectedColumn { name: "hll_value", column_types: &[ColumnType::Text], raw_type: None },
            ExpectedColumn { name: "percentile_value", column_types: &[ColumnType::Binary], raw_type: None },
        ],
    );
    assert_eq!(sketch_result.value(0, 0)?, Value::Null);
    assert_eq!(sketch_result.value(0, 1)?, Value::Null);
    assert_eq!(sketch_result.value(0, 2)?, Value::Null);
    sketch_result.close()?;

    Ok(())
}

fn assert_contains_text(value: Value, needle: &str) {
    match value {
        Value::Text(text) => assert!(text.contains(needle), "expected `{text}` to contain `{needle}`"),
        other => panic!("expected text value, got {other:?}"),
    }
}

fn assert_matches_text_prefix(value: Value, prefix: &str) {
    match value {
        Value::Text(text) => assert!(text.starts_with(prefix), "expected `{text}` to start with `{prefix}`"),
        other => panic!("expected text value, got {other:?}"),
    }
}

fn is_runtime_unavailable(error: &aq_database::Error) -> bool {
    let message = error.to_string();
    message.contains("Could not load")
        || message.contains("Library not loaded")
        || message.contains("connection refused")
    || message.contains("timed out")
}